node/avn-service/src/web3_utils.rs

```diff
@@ -0,0 +1,192 @@
+
+    let nonce = web3_data.get_nonce(sender_eth_address, false).await?;
+    let web3 = web3_data.get_web3_instance()?;
+    let gas_estimate =
+        estimate_gas(web3, sender_eth_address, recipient, &send_request.data).await?;
+
+    Ok(TransactionParameters {
+        nonce: Some(nonce.into()),
+        to: Some(H160::from_slice(recipient)),
+        value: U256::zero(),
+        gas: gas_estimate,
+        gas_price: None,
+        data: web3::types::Bytes(send_request.data.clone()),
+        ..Default::default()
+    })
+}
+
+#[allow(dead_code)]
+pub async fn get_chain_id(web3: &Web3<Http>) -> anyhow::Result<u64> {
+    Ok(web3
+        .eth()
+        .chain_id()
+        .await
+        .with_context(|| "Error getting chain Id".to_string())?
+        .as_u64())
+}
+
+async fn estimate_gas(
+    web3: &Web3<Http>,
+    sender: &Vec<u8>,
+    recipient: &[u8],
+    data: &Vec<u8>,
+) -> anyhow::Result<U256> {
+    let call_request = CallRequest {
+        from: Some(H160::from_slice(&sender)),
+        to: Some(H160::from_slice(recipient)),
+        gas: None,
+        gas_price: None,
+        value: Some(U256::zero()),
+        data: Some(Bytes(data.to_vec())),
+        access_list: None,
+        max_fee_per_gas: None,
+        max_priority_fee_per_gas: None,
+        transaction_type: None,
+    };
+
+    Ok(web3.eth().estimate_gas(call_request.clone(), None).await.with_context(|| {
+        format!(
+            "Error estimating gas for data: {}",
+            serde_json::to_string_pretty(&call_request).unwrap_or(format!("{:?}", call_request))
+        )
+    })?)
+}
+
+pub async fn get_current_block_number(web3: &Web3<Http>) -> anyhow::Result<u64> {
+    Ok(web3.eth().block_number().await?.as_u64())
+}
+
+pub async fn get_tx_receipt(
+    web3: &Web3<Http>,
+    tx_hash: ethereum_types::H256,
+) -> anyhow::Result<Option<TransactionReceipt>> {
+    Ok(web3.eth().transaction_receipt(web3::types::H256(tx_hash.0)).await?)
+}
+
+pub async fn send_raw_transaction(
+    web3: &Web3<Http>,
+    tx: Bytes,
+) -> anyhow::Result<web3::types::H256> {
+    Ok(web3
+        .eth()
+        .send_raw_transaction(tx)
+        .await
+        .with_context(|| format!("Error while sending raw transaction to Ethereum"))?)
+}
+
+// Based and refactored from: https://github.com/tomusdrw/rust-web3/blob/v0.18.0/src/signing.rs#L151-L172
+
+/// Gets the address of a public key.
+///
+/// The public address is defined as the low 20 bytes of the keccak hash of
+/// the public key. Note that the public key returned from the `secp256k1`
+/// crate is 65 bytes long, that is because it is prefixed by `0x04` to
+/// indicate an uncompressed public key; this first byte is ignored when
+/// computing the hash.
+pub fn public_key_address(public_key: &PublicKey) -> Address {
+    let public_key = public_key.serialize_uncompressed();
+
+    debug_assert_eq!(public_key[0], 0x04);
+    let hash = keccak256(&public_key[1..]);
+
+    Address::from_slice(&hash[12..])
+}
```

Comment:
I think this area could improve if we replace the magic numbers with const variables.